### PR TITLE
add better title parser

### PR DESCRIPTION
**Pull request by adedomin**
Can't actually test if it works because I don't have python2. do whatever.

### CODE REVIEW

**Comment by ineeee** [comment]
what does this fix?

edit: i mean, doesn't taiga output urls fine right now?

**Comment by ineeee** [comment]
i tested it and now it doesnt output anything when someone posts a link

i can't tell what's wrong, but heres stdout when posting `https://news.ycombinator.com/` in the chat

```
Traceback (most recent call last):
  File "core/reload.py", line 90, in reload
    eval(code, namespace)
  File "plugins/urls.py", line 10, in <module>
    from util.title_parse import tokenize_title_chunk, parse_title_chunk
  File "plugins/util/title_parse.py", line 56
    TitleTokenizer = re.compile(f'''(?x)
 (?P<{TitleParserTokens.WHITESPACE.name}>\\s+)
|(?P<{TitleParserTokens.TITLE.name}>[<][/]?\\s*[tT][iI][tT][lL][eE][^>]*[>])
|(?P<{TitleParserTokens.WORDS.name}>[^<>&"'\\s]+)
|(?P<{TitleParserTokens.DEC_CHAR_ENT.name}>
  [&][#]
   (?P<__internal_dec_code_value>\\d+)
  ;
 )
|(?P<{TitleParserTokens.HEX_CHAR_ENT.name}>
  [&][#]x
   (?P<__internal_hex_code_value>[a-fA-F0-9]+)
  ;
 )
|(?P<{TitleParserTokens.WORD_CHAR_ENT.name}>
  [&]
   (?P<__internal_word_code_value>\\w+;)
 )
|(?P<{TitleParserTokens.OTHER_TAG.name}>
  [<][/]?\\s*
   (?P<__internal_tag_name>[\\w!]+)
  [^>]*[>])
|(?P<{TitleParserTokens.INVALID.name}>.)
''')

  ^
SyntaxError: invalid syntax
```

**Comment by adedomin** [comment]
ok I fixed it for python2

File: plugins/util/title_parse.py [hunk 1]
```diff
@@ -113,12 +106,2253 @@ def parse_title_chunk(tokens):
...
+        try:
+            ret = ret + x
+        except UnicodeDecodeError:
+            ret = ret + x.decode('utf-8', 'ignore')
 
+    return ret
+
+
+# See: https://github.com/python/cpython/blob/3.8/Lib/html/entities.py#L264
+char_entities = {
```
Comment (ineeee) [inline plugins/util/title_parse.py:120]: i dont like this

**Comment by ineeee** [comment]
i dont like the 2358 lines, but hey it solves the issue. the bold/underline/italic support is nice, too.

up to u @FrozenPigs

**Review by ineeee — APPROVED** [review]
wtf do i put on this small comment box

edit: oh lol

**Review by ineeee — CHANGES_REQUESTED** [review]
i can't accept a bump from 3.3 to 4.5 without extensive testing. lxml is used *everywhere*

**Comment by adedomin** [comment]
it's more work than I feared, bot just keeps getting more problematic the more I probe it. outputs broken extended-ascii or some latin-1 encoding anyhow unless controlling pty is sufficient enough for the python interpreter. not sure how this affects a literal socket, but it seems to. python2 is just too broken and this is more pissing me off than it's worth.

**Pull request closed without merging**